--- include/CL/cl.hpp.in	2014-03-17 17:47:57.147223265 +0200
+++ include/CL/cl.hpp	2014-03-17 17:55:52.620234165 +0200
@@ -214,8 +214,6 @@
 #if defined(linux) || defined(__APPLE__) || defined(__MACOSX)
 #include <alloca.h>
 
-#include <emmintrin.h>
-#include <xmmintrin.h>
 #endif // linux
 
 #include <cstring>
@@ -1038,8 +1036,14 @@
 #endif // !_WIN32
     }
 
-    inline void fence() { _mm_mfence(); }
-}; // namespace detail
+    inline void fence() { 
+#ifdef _MSC_VER
+        _mm_mfence();
+#else
+        __sync_synchronize();
+#endif
+    }
+} // namespace detailssss
 
     
 /*! \brief class used to interface between C++ and
