--- a/src/utilities/aslParametersManager.cxx	2017-04-11 11:59:40.208368454 +0200
+++ b/src/utilities/aslParametersManager.cxx	2017-04-11 12:00:34.676369980 +0200
@@ -332,7 +332,7 @@
 
 		try
 		{
-			ifstream ifs(paramFile);
+			std::ifstream ifs(paramFile);
 			if (!ifs.good())
 				errorMessage("Can not open parameters file: " + paramFile);
 
@@ -357,7 +357,7 @@
 
 	void ParametersManager::writeParametersFile(const std::string fileName)
 	{
-		ofstream fo(fileName);
+		std::ofstream fo(fileName);
 		if (!fo.good())
 			errorMessage("ParametersManager::writeParametersFile() - can not open file: " + fileName);
 
@@ -463,7 +463,7 @@
 				}
 				else
 				{
-					ifstream ifs(p.string());
+					std::ifstream ifs(p.string());
 					if (ifs.good())
 					{
 						parsed_options parsed = parse_config_file(ifs, allOptions, true);
