diff --git a/samples/CMakeLists.txt b/samples/CMakeLists.txt
index 502ee6d..c722a97 100644
--- a/samples/CMakeLists.txt
+++ b/samples/CMakeLists.txt
@@ -53,5 +53,5 @@ add_subdirectory( Ch14_common_parallel_patterns )
 add_subdirectory( Ch15_gpus )
 add_subdirectory( Ch16_cpus )
 add_subdirectory( Ch17_fpgas )
-add_subdirectory( Ch18_using_libs )
+# add_subdirectory( Ch18_using_libs )
 add_subdirectory( Ch19_memory_model_and_atomics )
